src/mount.rs

```diff
@@ -0,0 +1,655 @@
+use lru::LruCache;
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use std::{
+    borrow::Cow,
+    cmp::Ordering,
+    ffi::OsStr,
+    fs,
+    io::{self, ErrorKind, Read},
+    iter::Peekable,
+    num::NonZeroUsize,
+    ops::ControlFlow,
+    path::{Component, Path, PathBuf},
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+use tracing::{error, info, trace};
+
+use bytes::Bytes;
+use itertools::Itertools;
+use tracing::{debug, warn};
+use windows_projfs::{
+    DirectoryEntry, DirectoryInfo, FileInfo, Notification, ProjectedFileSystem,
+    ProjectedFileSystemSource,
+};
+
+use crate::{
+    counters::Counter,
+    monitor::{
+        task::{Task, TaskList},
+        Monitor, Problem,
+    },
+    Apath, Archive, BandId, BandSelectionPolicy, IndexEntry, IndexRead, Kind, Result, StoredTree,
+};
+
+macro_rules! static_dir {
+    ($name:expr) => {
+        DirectoryInfo {
+            directory_name: ($name).to_string(),
+            directory_attributes: DIRECTORY_ATTRIBUTES,
+
+            ..Default::default()
+        }
+        .into()
+    };
+}
+
+struct VoidMonitor;
+impl Monitor for VoidMonitor {
+    fn count(&self, _counter: Counter, _increment: usize) {}
+
+    fn set_counter(&self, _counter: Counter, _value: usize) {}
+
+    fn problem(&self, _problem: Problem) {}
+
+    fn start_task(&self, name: String) -> Task {
+        /*
+         * All data related to the target task will be dropped
+         * as soon the callee drops the task.
+         */
+        let mut list = TaskList::default();
+        list.start_task(name)
+    }
+}
+
+#[derive(Debug)]
+struct HunkMetaInfo {
+    index: u32,
+
+    start_path: Apath,
+    end_path: Apath,
+}
+
+struct HunkHelper {
+    hunks: Vec<HunkMetaInfo>,
+}
+
+impl HunkHelper {
+    pub fn from_index(index: &IndexRead) -> Result<Self> {
+        let mut hunk_info = index
+            .hunks_available()?
+            .into_par_iter()
+            .map(move |hunk_index| {
+                let mut index = index.duplicate();
+                let entries = index.read_hunk(hunk_index)?;
+                let meta_info = if let Some(entries) = entries {
```

Comment:
This seems to read and then discard every hunk, and I wonder if it would be better to cache them? 

If the tree is very large that might use a lot of memory, but it seems like you at least might as well put them in an LRU?

Reply:
Caching the hunk contents for the current implementation of the HunkHelper itself does not make sense, as it never needs any of the hunk contents.  
The moment we actually access the hunk contents later on via `load_hunk_contents` the results are getting cached.  
  
Another logical issue with using an cache at this stage is, that the `HunkHelper` loads every hunk once, therefore we do not gain any knowledge about which hunk is most likely be loaded again.